lib/datadog/tracing/contrib/aws/service/states.rb

```diff
@@ -16,21 +16,35 @@ def add_tags(span, params)
               execution_arn = params[:execution_arn]
               state_machine_account_id = ''
 
+              if state_machine_arn
+                span.set_tag(Aws::Ext::TAG_STATE_MACHINE_ARN, state_machine_arn)
+                # https://docs.aws.amazon.com/step-functions/latest/apireference/API_StartExecution.html#API_StartExecution_RequestSyntax:~:text=Required%3A%20No-,stateMachineArn,-The%20Amazon%20Resource
+                # arn:<partition>:states:<region>:<account-id>:stateMachine:<myStateMachineName>
+                # arn:<partition>:states:<region>:<account-id>:stateMachine:<myStateMachineName>:10
+                # arn:<partition>:states:<region>:<account-id>:stateMachine:<myStateMachineName:PROD>
+                # There are 3 patterns to cover and attempt to capture the `myStateMachineName`, it should always be in index 6 and account_id at index 4
+                parts = state_machine_arn.split(':')
+                state_machine_name ||= parts[6]
+                state_machine_account_id = parts[4]
+              end
+
               if execution_arn
```

Comment:
The order of this seemed to flow better as the state_machine_arn is easier to understand